ui/src/logs/utils/index.ts

```diff
@@ -174,33 +176,125 @@ export function buildGeneralLogQuery(
   return `${select}${condition}${dimensions}${fillClause}`
 }
 
-export function buildBackwardLogQuery(
+export async function queryCount(
+  lower,
+  upper,
+  config,
+  filters,
+  searchTerm,
+  proxyLink,
+  namespace
+): Promise<number> {
+  const {database, retentionPolicy, measurement} = config
+  const rpSegment = retentionPolicy ? `"${retentionPolicy}"` : ''
+  const fullyQualifiedMeasurement = `"${database}".${rpSegment}."${measurement}"`
+  const select = `SELECT count(message) FROM ${fullyQualifiedMeasurement}`
+  let condition = `WHERE time >= '${lower}' AND time <='${upper}'`
+
+  if (!_.isEmpty(searchTerm)) {
+    condition = `${condition} AND message =~ ${new RegExp(searchTerm)}`
+  }
+
+  if (!_.isEmpty(filters)) {
+    condition = `${condition} AND ${filtersClause(filters)}`
+  }
+
+  const query = `${select} ${condition} FILL(0)`
+  const result = await executeQueryAsync(proxyLink, namespace, query)
+  return getDeep<number>(result, 'results.0.series.0.values.0.1', 0)
+}
+
+export async function findBackwardLower(
   upper: string,
   config: QueryConfig,
   filters: Filter[],
-  searchTerm: string | null = null
-) {
-  const {tags, areTagsAccepted} = config
+  searchTerm: string | null = null,
+  proxyLink: string,
+  namespace: Namespace
+): Promise<string> {
+  const parsedUpper = moment(upper)
+
+  let secondsBack = 30
+  let currentLower = parsedUpper.subtract(secondsBack, 'seconds')
+
+  while (true) {
+    if (secondsBack > SECONDS_AWAY_LIMIT) {
+      // One day
+      break
+    }
 
-  const condition = buildInfiniteWhereClause({
-    upper,
-    tags,
-    areTagsAccepted,
-  })
+    const count = await queryCount(
+      currentLower.toISOString(),
+      upper,
+      config,
+      filters,
+      searchTerm,
+      proxyLink,
+      namespace
+    )
+
+    if (count >= 400) {
+      break
+    }
 
-  return buildGeneralLogQuery(condition, config, filters, searchTerm)
+    secondsBack *= secondsBack
```

Comment:
would you consider adding a code comment to provide context as to why this choice is being made?